Chads branch

chadlyalan opened the pull request:
## Description
Fixes for [Issue 1](https://github.com/chadlyalan/c932a/issues/1)

The first part was figuring out how to get the updated data from the store whenever it changed.

At first there was an error every time I added a new message about the conversationId. I found out that the postMessage function was calling an asynchronous function, and putting it in a normal const variable. The first change I had to do was make that await. From then on the "data" variable was actually usable, whereas before it wasn't readable by the rest of the components.

Then I added a useSelector to the ActiveChat component so that the message data that ActiveChat was sending to Messages was always going to be the updated data from the store, as that is updated.

Then no matter where the message gets updated, if the active chat is displaying that conversation, the UI will update with the new message.


The second part was fixing the order in which the messages displayed in the UI. I couldn't just reverse the order of everything (I tried that first) because then new messages were displayed in the opposite order as well.

I had to mutate the order in which Messages were stored in their conversations right after they were fetched. I added a new function to ThunkConversations that was called by fetchConversations so each messages array within all of the conversations was reversed, showing oldest first, then newest last. That way when messages displayed messages in order, it was the correct order with older messages on top of the page, and the latest messages at the bottom (closer to the input field).

I've also removed the db.js from main, and added it to the gitignore so hopefully the cypress tests can run as designed. 

## Notes on your approach and thought process

The biggest hurdle for me was understanding the flow of data. I'm fairly comfortable with react, but I am brand new to redux, and working with a database on the backend.

Most of my time was dedicated to reading redux docs, and comparing them to the starter code so I could understand what was happening.

I added logs to better understand everything better, and once I did I was able to pick apart the issues pretty quickly.

## Further comments (optional)

chadlyalan committed:
logging and fixing postMessage

As I've been learning more about hooks and redux I've adding some additional console logging to better understand the flow of data.

The postMessage function inside thunkCreators.js was async as was saveMessage, the function it calls, but the return value wasn't await so I added that.

After doing that I fixed how the reducer function receives that data. Now the correct data goes to addMessageToStore.

chadlyalan committed:
ActiveChat updates UI with changes

I added a useSelector to activeChat and now my conversations UI will update when there are changes to the conversations store.
I also accidentally broke addNewMessageToStore and fixed that back as well.

chadlyalan committed:
Fixing the order of messages

Now the Messages in each conversation have been re-ordered so that the oldest is on top, and the latest is on bottom (closer to the input bar).

I did this by changing the messages array within the greater data object, then I sent that new version of data to the dispatch just like the previous one.

chadlyalan committed:
small update

fixing some changes I made to sendMessage when I didn't understand what was going on.

chadlyalan committed:
adding db.js to the gitignore

db.js shouldn't be pushed up to main anymore, and I removed db.js from main as well. This commit should allow the tests to run successfully.

Tahoum reviewed (commented):
Thank you for your work on this PR 👍🏽 
Please address the comments below and let us know by using the "/hatchy pr" command on Slack and another reviewer will come by to give this PR another look and give you approval 🙂

chadlyalan committed:
PR response changes

Cleaning up some logging and my reverseMessages function as suggested by Tahoum

chadlyalan committed:
Update db.js

hiding my credentials through the secret .env file

chadlyalan committed:
making reducer functions immutable

I modified addMessageToStore and addNewConvoToStore so that the return a copy of the conversations to make an immutable change to state rather than mutating it directly.

This allows me to remove the useSelector in ActiveChat, and for the ActiveChat component to still update when changes to the store are made.

chadlyalan committed:
Delete db.js

chadlyalan committed:
trying to make sure db.js doesn't get pushed to github

chadlyalan committed:
Merge branch 'chadsBranch' of https://github.com/chadlyalan/c932a into chadsBranch

chadlyalan committed:
Update .gitignore

chadlyalan committed:
adding the db.js

I'm removing the db.js from the gitignore so that the tests can run
successfully, but this version uses environmental variables so the credentials to the database are hidden to the public.

chadlyalan committed:
Update db.js

The Tests are still failing on Cypress. I see errors about the connection being refused at 127.0.0.1:3000. I'm thinking that might not be a smart way for me to connect to the database since it is possible to be running on a different local IP. Using Localhost seems like a more universal option.

chadlyalan committed:
update to db.js environment variables

I changed the Sequelize connection to the postgres database to be a string literal, and correctly use the environment variables from my .env file.

I also removed the useSelector import that wasn't being used in ActiveChat anymore.

chadlyalan committed:
dynamic rendering of image

I have begun to style both of my welcome pages and have gone through different methods of styling. I am now working on styling dynamically so small (mobile) screens also look good.

Currently the picture that displays on the side of the screen only renders when the size is at least "medium"

chadlyalan committed:
Update Signup.js

updating some variable names to help readability

chadlyalan committed:
dynamic screen sizing

The main two grid items now resize themselves correctly as the screen's size changes.

chadlyalan committed:
refactor of signup

I moved most of the component rendering away from signup.js so that I could conditionally render based on screen size in a more clean way.
I can now re-style and have more accurate useStyles for each component, even if it's the same component, but for a different screen size.

chadlyalan committed:
styling

cleaning up some duplicate styling being applied to the same elements in two different files.

chadlyalan committed:
Fixing structure

After doing some research on best practices I came up with a different method to dynamically change the layout of the grid, and make sure that I can still have a landing image on all versions that looks good.
There are still a few things that aren't quite right with the styling but it's a lot better than it was last commit.

chadlyalan committed:
Update Signup.js

Mostly finished product for Signup.js
(The registration landing page)

chadlyalan committed:
Update Signup.js

Signup page is almost done. We're just missing the chat bubble icon.

chadlyalan committed:
final push for styling of landing pages

I extracted the styling from both of the pages to external style sheets that use makeStyles and then applied both of them to both of the landing pages: register, and login.

Now the pages should have decent styling and is updated based on screen size.
